uwb-beacon-firmware/src/state_estimation_thread.cpp

```diff
@@ -0,0 +1,127 @@
+#include <ch.h>
+#include <hal.h>
+#include "main.h"
+#include "state_estimation.hpp"
+#include "ranging_thread.h"
+#include "state_estimation_thread.h"
+#include "anchor_position_cache.h"
+#include "imu_thread.h"
+
+/** Parameters for this service. */
+static struct {
+    parameter_namespace_t ns;
+    parameter_t process_variance;
+    parameter_t range_variance;
+} params;
+
+/** Creates the parameters. */
+static void parameters_init(parameter_namespace_t *parent);
+
+static THD_WORKING_AREA(state_estimation_wa, 1024);
+static THD_FUNCTION(state_estimation_thd, arg)
+{
+    (void) arg;
+
+    messagebus_topic_t *range_topic, *imu_topic;
+    struct {
+        mutex_t lock;
+        condition_variable_t cv;
+        messagebus_watchgroup_t group;
+        messagebus_watcher_t watchers[2];
+    } watchgroup;
+
+    messagebus_topic_t state_estimation_topic;
+    MUTEX_DECL(state_estimation_topic_lock);
+    CONDVAR_DECL(state_estimation_topic_condvar);
+    position_estimation_msg_t state_estimation_topic_content;
+
+    messagebus_topic_init(&state_estimation_topic,
+                          &state_estimation_topic_lock,
+                          &state_estimation_topic_condvar, &state_estimation_topic_content,
+                          sizeof(state_estimation_topic_content));
+    messagebus_advertise_topic(&bus, &state_estimation_topic, "/ekf/state");
+
+    parameters_init(&parameter_root);
+
+    // TODO configure variance
+    RadioPositionEstimator estimator;
+
+    range_topic = messagebus_find_topic_blocking(&bus, "/range");
+    imu_topic = messagebus_find_topic_blocking(&bus, "/imu");
+
+    /* Prepare to listen on all groups. */
+    chMtxObjectInit(&watchgroup.lock);
+    chCondObjectInit(&watchgroup.cv);
+    messagebus_watchgroup_init(&watchgroup.group, &watchgroup.lock, &watchgroup.cv);
+    messagebus_watchgroup_watch(&watchgroup.watchers[0],
+                                &watchgroup.group,
+                                range_topic);
+    messagebus_watchgroup_watch(&watchgroup.watchers[1],
+                                &watchgroup.group,
+                                imu_topic);
+
+    while (true) {
+        messagebus_topic_t *topic;
+        topic = messagebus_watchgroup_wait(&watchgroup.group);
+
+        if (topic == range_topic) {
+            // If we got a range, feed it to the estimator
+            range_msg_t msg;
+            anchor_position_msg_t *anchor_pos;
+            messagebus_topic_read(topic, &msg, sizeof(msg));
+
+            // Discard messages with range greater than 1km, as they are
+            // probably the result of an underflow if the tag is too close to
+            // the anchor.
```

Comment:
lol wut?

Reply:
basically if you put 2 tags within like 5cm of each other, when computing the distance you get an underflow with an ill-calibrated system. was just easier to filter them.

Reply:
Fair enough